### ISSUE TITLE
Pyright in "strict" mode shows warnings for overloads not used in client code

### ISSUE THREAD

**Issue by AlexanderPodorov** [microsoft/pyright#4777]
**Describe the bug**
Pyright in "strict" mode shows warnings for overloads not used in client code.
The example could be the `types-aioboto3` package which provide typing information for `aioboto3` library.
As a more common example, `boto3-stubs` package provides type hints for `boto3` library.
The way how those type stubs work is that a client (calling code) should define which service it wants to use, essentially installing e.g. `types-aioboto3[s3, dynamodb]` if it wants S3 storage or DynamoDB type hints only.
The way how `types-aioboto3` or `boto3-stubs` works is that they generate type stubs and for some reason include missing imports for unused services. That's not perfect of course, however pyright could not raise a warning in case if type for selected overload is known.
See this issue for all of details: https://github.com/youtype/mypy_boto3_builder/issues/180

**To Reproduce**
```python
from aioboto3 import Session
from types_aiobotocore_s3.client import S3Client

def create_s3_client() -> S3Client:
        return Session().client(
            service_name='s3',  # respective Client class is defined by this field, essentially via typing.overload
            endpoint_url='some url',
            aws_access_key_id='***',
            aws_secret_access_key='***'
        )
```
Type for 's3' overload is known to be S3Client, however pyright shows warnings for other services which are not used.
See this particular comment: https://github.com/youtype/mypy_boto3_builder/issues/180#issuecomment-1466830928

3rd party packages involved:
```
aioboto3
types-aioboto3[s3]
```

**Expected behavior**
No warnings. As reported by the `types-aioboto3` maintainer, no warnings are thrown if `types-aioboto3[all]` is installed, however it's not an option, since it will install ~340 unused packages.

**VS Code extension or command-line**
Pyright is run via pylance extension in VSCode, version is `2023.3.20`.

**Additional context**
See: https://github.com/youtype/mypy_boto3_builder/issues/180

**Comment by erictraut** [comment]
Pyright is working as designed here. The method `Session.client` is defined as an overload, and some of the overload signatures contain type annotations that refer to symbols whose types are unknown. If you enable the `reportUnknownMemberType` check in pyright (which is included in the `strict` ruleset), pyright will report the case where you are referencing a member (an attribute or method) whose type is partially unknown. 

In your example, the error is detected with the expression `Session().client` because `client` is a member whose type is partially known. This expression evaluates to an overloaded function, and once that function is called, the result may no longer be Unknown (depending on which packages you have installed and which overload is selected).

Overload selection is complicated and sometimes requires evaluation of the return types of signatures as well as the input parameter types. If the return types are unknown, it can affect overload selection in some cases, so it makes sense for pyright to notify you of this potential "hole" in type checking when you're using strict mode.

**Comment by AlexanderPodorov** [comment]
Thanks for the explanation! Yes, I agree with you.
I just wanted to mention that missing imports/missing types are done on purpose in this case, this is how `types-aioboto3` works.
This particular overload with `service_name='s3'` is known and there should be no problem to use it.
In case if I define other `service_name` for which I don't have type stubs, there should be a warning thrown, that's right.
In case if we are able to match a particular overload (in this case it is matched by `service_name` `Literal`) and all types here are known, does it makes sense to throw a warning? Thanks!

**Comment by erictraut** [comment]
As I mentioned, the error is detected on the member access (dot) expression `Session().client`, not on the call expression `Session().client(...)`. Overloads are evaluated as part of the call expression. The member access expression is evaluated (and any errors reported) long before any overloads are considered.

**Comment by AlexanderPodorov** [comment]
Thanks for clarification, I'm closing this issue.
